### PR TITLE
Use threading for more precise time intervals

### PR DESCRIPTION
**Pull request by palto42**
The current while loop doesn't account for the duration to get the readings from the inverter, so that the actual time interval is always a bit longer that specified (in my case about 0.8s).
I propose to replace the while loop with a simple threading class which ensures that the time interval will not slip over time.

### CODE REVIEW

**Comment by kbialek** [comment]
@palto42 What real-life problem does it solve? In what use case such high interval accuracy is required?

**Comment by palto42** [comment]
@kbialek I agree that ~0.8 seconds drift per measurement may not cause any significant issue, but at least I was a bit surprised seeing it in my measurement time series. If you run the measurements at shorter intervals like 10 s, then it will be more relevant.
The solution isn't too complex in my view and only uses standard Python libraries, so I though it's worth to suggest it. But if you think it's an overkill, it's ok for me if you reject it.

File: deye_daemon.py [hunk 1]
```diff
@@ -90,12 +91,29 @@ def __remove_duplicated_reg_ranges(self, reg_ranges: list[SensorRegisterRange])
         return result
 
 
+class setInterval :
```
Comment (kbialek) [inline deye_daemon.py:94]: Please rename this class to something more meaningful, and start the classname with uppercased letter.

File: deye_daemon.py [hunk 2]
```diff
@@ -90,12 +91,29 @@ def __remove_duplicated_reg_ranges(self, reg_ranges: list[SensorRegisterRange])
...
+        self.action=action
+        self.stopEvent=threading.Event()
+        thread=threading.Thread(target=self.__setInterval)
+        thread.start()
+
+    def __setInterval(self) :
+        nextTime=time.time()+self.interval
+        while not self.stopEvent.wait(nextTime-time.time()) :
+            nextTime+=self.interval
+            self.action()
```
Comment (kbialek) [inline deye_daemon.py:106]: What will happen with `action()` throws an exception?

File: deye_daemon.py [hunk 3]
```diff
@@ -90,12 +91,29 @@ def __remove_duplicated_reg_ranges(self, reg_ranges: list[SensorRegisterRange])
...
+
+    def cancel(self, _signum, _frame):
+        self.stopEvent.set()
```
Comment (kbialek) [inline deye_daemon.py:109]: Please format this code - at least add spaces. There is no project wide code formatter yet, so be creative. Use black or whatever you prefer.

**Review by kbialek — CHANGES_REQUESTED** [review]
Let's improve the code before merging it

File: deye_daemon.py (hunk 3 above)
Comment (palto42) [reply deye_daemon.py:109]: Sorry, usually using black but forgot it here, fixed.

File: deye_daemon.py (hunk 2 above)
Comment (palto42) [reply deye_daemon.py:106]: Added exception handler with logging

File: deye_daemon.py (hunk 1 above)
Comment (palto42) [reply deye_daemon.py:94]: Renamed the class to `IntervalRunner`.

**Comment by palto42** [comment]
@kbialek I updated the code as requested and rebased the branch on latest main.

File: deye_daemon.py [hunk 4]
```diff
@@ -90,12 +93,35 @@ def __remove_duplicated_reg_ranges(self, reg_ranges: list[SensorRegisterRange])
...
+        thread.start()
+
+    def __setInterval(self):
+        nextTime = time.time() + self.interval
+        while not self.stopEvent.wait(nextTime - time.time()):
+            nextTime += self.interval
+            try:
+                self.action()
+            except Exception as e:
+                self.__log.exception("Unexpected error during daemon execution")
```
Comment (kbialek) [inline deye_daemon.py:113]: @palto42 Could you also pass `e` to the exception method? This way it will be more visible where it crashed.

**Review by kbialek — CHANGES_REQUESTED** [review]
Looks good. Just one minor remark.

File: deye_daemon.py (hunk 4 above)
Comment (palto42) [reply deye_daemon.py:113]: @kbialek It is actually not required to pass `e` in `logging.exception`, the logger automatically adds it to the message, see [loggin.exception](https://docs.python.org/3.10/library/logging.html#logging.exception).

Actually I found that the way you use it in `deye_connector.py` is incorrect and causes an exception in the logger:

Example:

```python
❯ python
Python 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import logging
>>> try:
...     1/0
... except Exception:
...     logging.exception("Something unexpected happened")
... 
ERROR:root:Something unexpected happened
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ZeroDivisionError: division by zero
>>> 
>>>
>>> try:
...     1/0
... except Exception as e:
...     logging.exception("Something unexpected happened: %s", e)
... 
ERROR:root:Something unexpected happened: division by zero
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ZeroDivisionError: division by zero
>>>
```
Actually I found that the way you use it in `deye_connector.py` is wrong and causes an exception in the logger:

```python
>>> try:
...     1/0
... except Exception as e:
...     logging.exception("Something unexpected happened", e)
... 
--- Logging error ---
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ZeroDivisionError: division by zero

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/logging/__init__.py", line 1100, in emit
    msg = self.format(record)
  File "/usr/lib/python3.10/logging/__init__.py", line 943, in format
    return fmt.format(record)
  File "/usr/lib/python3.10/logging/__init__.py", line 678, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.10/logging/__init__.py", line 368, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "<stdin>", line 4, in <module>
Message: 'Something unexpected happened'
Arguments: (ZeroDivisionError('division by zero'),)
>>> 
```

I fixed the exception logging in both `deye_connector.py` and `deye_daemon.py` now.

**Comment by kbialek** [comment]
Nice catch! That is what happens when Java programmer starts coding in Python :smile:

**Review by kbialek — APPROVED** [review]

**Pull request merged**